bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserSpigot1_11CraftingListener.java

```diff
@@ -0,0 +1,196 @@
+                        Collections.singletonList(output), uuid, "crafting_table", 0, netId++));
+            } else if (recipe instanceof ShapelessRecipe) {
+                ShapelessRecipe shapelessRecipe = (ShapelessRecipe) recipe;
+                Ingredient[] ingredients = new Ingredient[shapelessRecipe.getIngredientList().size()];
+                ItemData[] input = new ItemData[shapelessRecipe.getIngredientList().size()];
+                for (int i = 0; i < input.length; i++) {
+                    Pair<ItemStack, ItemData> result = translateToBedrock(session, shapelessRecipe.getIngredientList().get(i));
+                    ingredients[i] = new Ingredient(new ItemStack[]{result.getKey()});
+                    input[i] = result.getValue();
+                    isNotAllAir = isNotAllAir || input[i].getId() != 0;
+                }
+                if (!isNotAllAir) continue;
+                UUID uuid = UUID.randomUUID();
+                // Add recipe to our internal cache
+                ShapelessRecipeData data = new ShapelessRecipeData("", ingredients, javaOutput);
+                session.getCraftingRecipes().put(netId,
+                        new com.github.steveice10.mc.protocol.data.game.recipe.Recipe(RecipeType.CRAFTING_SHAPELESS, uuid.toString(), data));
+                // Add recipe for Bedrock
+                craftingDataPacket.getCraftingData().add(CraftingData.fromShapeless(uuid.toString(),
+                        Arrays.asList(input), Collections.singletonList(output), uuid, "crafting_table", 0, netId++));
+            }
+        }
+
+        session.sendUpstreamPacket(craftingDataPacket);
+    }
+
+    @SuppressWarnings("deprecation")
+    private Pair<ItemStack, ItemData> translateToBedrock(GeyserSession session, org.bukkit.inventory.ItemStack itemStack) {
+        if (itemStack != null && itemStack.getData() != null) {
+            if (itemStack.getType().getId() == 0) {
+                return new Pair<>(null, ItemData.AIR);
+            }
+            int legacyId = (itemStack.getType().getId() << 4) | (itemStack.getData().getData() & 0xFFFF);
+            if (itemStack.getType().getId() == 355 && itemStack.getData().getData() == (byte) 0) { // Handle bed color since the server will always be pre-1.12
+                legacyId = (itemStack.getType().getId() << 4) | ((byte) 14 & 0xFFFF);
+            }
+            // old version -> 1.13 -> 1.13.1 -> 1.14 -> 1.15 -> 1.16 and so on
+            int itemId;
+            if (mappingData1_12to1_13.getItemMappings().containsKey(legacyId)) {
+                itemId = mappingData1_12to1_13.getNewItemId(legacyId);
+            } else if (mappingData1_12to1_13.getItemMappings().containsKey((itemStack.getType().getId() << 4) | (0))) {
+                itemId = mappingData1_12to1_13.getNewItemId((itemStack.getType().getId() << 4) | (0));
+            } else {
+                // No ID found, just send back air
+                return new Pair<>(null, ItemData.AIR);
```

Comment:
The following code is quite compact. I personally put readability above line count.